### ISSUE TITLE
Can "namespace imports" be removed?

### ISSUE THREAD

**Issue by jplaisted** [google/closure-compiler#2767]
Can "namespace imports" (e.g. `import useragent from 'goog:goog.useragent'`) be removed? Is anyone using these externally?

These are becoming a bit of a burden to support since code for them is not centralized and seems to have been added a bit ad hoc. Also they don't make much sense since the browser cannot load them and they have different semantics than important an ES6 module ("destructures" don't update). I'd rather people just use `goog.require` within an ES6 module if they want a Closure file.

+@ChadKillingsworth

**Comment by jplaisted** [comment]
The expected migration would be to stop using `import` and use `goog.require` as you would in a non-ES6 module. e.g. `import { f } from 'goog:namespace'` becomes `const {f} = goog.require('namespace')`

**Comment by ChadKillingsworth** [comment]
The Angular team added them - @mprobst I believe. https://groups.google.com/d/msg/closure-compiler-discuss/fpE8QnTtacQ/FnaiJDMbCQAJ

**Comment by jplaisted** [comment]
fwiw these are also broken for non-legacy `goog.module`s. [Example](https://closure-compiler-debugger.appspot.com/#input0%3Dgoog.module('my.module')%253B%250A%250Aexports.foo%2520%253D%25200%253B%250Aexports.bar%2520%253D%25201%253B%26input1%3Dimport%2520MyModule%2520from%2520'goog%253Amy.module'%253B%250A%250Ause(MyModule)%253B%250Ause(MyModule.foo)%253B%250Ause(MyModule.bar)%253B%26conformanceConfig%26externs%3D%252F**%2520%2540param%2520%257B...%253F%257D%2520p%2520*%252F%250Afunction%2520use(p)%2520%257B%257D%26refasterjs-template%26CHECK_SYMBOLS%3Dtrue%26MISSING_PROPERTIES%3Dtrue%26TRANSPILE%3Dtrue%26CHECK_TYPES%3Dtrue%26CLOSURE_PASS%3Dtrue%26PRESERVE_TYPE_ANNOTATIONS%3Dtrue%26PRETTY_PRINT%3Dtrue).

**Comment by ChadKillingsworth** [comment]
@jplaisted What happens for legacy provides?

I think as long as all the use cases are covered, they could be deprecated. They are used - but not widely.

**Comment by mprobst** [comment]
@alexeagle might be using these with TS + Closure + ES6 modules? The g3 internal TypeScript integration resolves everything to `goog.require` (ES6 modules aren't quite working well enough for it), so it's unused from that angle.

But @concavelenz might want to take a look, using namespace based modules from ES6 modules is a problem that does need solving in some way, doesn't it?

**Comment by alexeagle** [comment]
I don't anticipate ABC (g.co/ng/abc) supporting a use case externally of a user mixing Closure-annotated JS sources with TypeScript sources, so I don't think I have a stake in this.

I agree that users ought to be able to adopt ES6 module syntax in place of `goog` and this feature seems like a necessary interop to allow such a migration to be incremental.

**Comment by jplaisted** [comment]
@ChadKillingsworth this style of import assumes the namespace exists, which it does for provides and legacy modules. But not for non-legacy goog.modules.

@alexeagle I don't think we allow having a `goog.provide` or `goog.module` in an ES6 module. imo changing `import 'goog:namespace'` to `import 'path'`isn't really different than changing `goog.require(namespace)` to `import 'path'`

Our planned migration flow would be to rename your Closure file to `.legacy.js` (or something), create a new ES6 module with the old name, migrate all of the logic into the ES6 module, and leave the Closure file as a small shell that just re-exports the new ES6 module (`goog.require` will be able to reference ES6 modules by path from `goog.module`s). Then you can update the imports/requires at your leisure. Example:

```
// namespace.legacy.js
goog.module('my.legacy.namespace');
goog.module.declareLegacyNamespace(); // optional
const ES6 = goog.require('./namespace.js');
exports = ES6;
```
```
// Assume we have this (in an ES6 module or a Closure file, doesn't matter):
const namespace = goog.require('my.legacy.namespace');
use(namespace.foo);

// We can migrate it to this whenever we want:
import * as namespace from './namespace.js';
use(namespace.foo);
```

**Comment by mprobst** [comment]
YMMV, but that'd mean somebody who wants to use `path/to/shared/library.js` from `my/application.js` now has to make modifications to the shared library to be able to use them from ES6 modules. That might be a problem, no?

**Comment by jplaisted** [comment]
Why would library modifications be needed?

**Comment by mprobst** [comment]
I might be misunderstanding you. If `path/to/shared/library.js` is a `goog.module` and you want to import it from ES6, you'd need to move it to `path/to/shared/library.legacy.js` and create an ES6 module at `path/to/shared/library.js` in its place that re-exports, wouldn't you? You could also create it in your app, but that seems odd.

**Comment by jplaisted** [comment]
If you want to use `import` yes. But there's no issue with using `goog.require` in an ES6 module. We don't allow `goog.provide` or `goog.module`, but `goog.require` is totally fine. Which is one reason why I'd like to get rid of `import 'goog:'`.

So if `path/to/shared/library.js` is a `goog.module` and you want to use it from an ES6 module just `goog.require` it.

**Comment by jplaisted** [comment]
I think there was some discussion that I was not apart of today that involved @mprobst and @shicks where there was an agreement this was okay to remove? At least once TypeScript stops emitting Javascript with it.

**Comment by jplaisted** [comment]
Okay so re-reading the thread it seems like no one is  using it but the argument is that is a preferred migration path, to which I completely disagree. `goog` is a global symbol and will be available in ES6 modules. And thus so will `goog.require`.

I argue users should call `goog.require` from ES6 modules to get references to `goog.provide`s/`goog.module`s.

I believe `import` importing non-ES6 modules is a bad pattern (and I think so for other module systems too, like CJS). `import` will have totally different semantics for ES6 modules and non-modules. At least with `goog:` it is a little clearer, in that you're **not** importing an ES6 module. But with CJS if you import a path; and you have no idea if what you're importing is an ES6 or CJS module or not (right?). But in either case you have no idea what your import statement is doing until you open the file you're importing due to the way CJS and `goog.module` handle default exports (`exports = value`). With ES6 modules `import *` is always an object that contains every export as a property; but this may not be the case with `goog.module` and CJS. Additionally what does a default import even mean for CJS/`goog.module`?

tl;dr ES6 import for non-ES6 modules is not well defined and will lead to developer confusion.

**Comment by ChadKillingsworth** [comment]
ES6 imports of CJS is being thoroughly defined by NodeJS right now. It's already very much defined by other transpilers. The interaction in Closure-Compiler is already defined and correct. The only thing missing is cross module type checking.

**Comment by mprobst** [comment]
@jplaisted sounds reasonable.

The drawback is that users writing in ES6 modules importing pre-module code don't get the nicer import syntax and have to deal with the inconsistency of two import systems across their code base.

In TS land, we though it'd be nicer for users to have one consistent system. Ideally we'd actually have imports pointing to the file that has the provide, not just the `goog:` prefixed namespace (that'd also make dependency management a lot easier), but that's of course tricky given that there's no easy mapping of provides to files.

In TS, we also needed to have proper ES6 imports as the compiler doesn't understand that `goog.require` brings in a value but also a type, so that was another limiting factor, so using `goog.require` just doesn't work.

But I understand that's a balance and a judgement call you need to make there. In any case, TS is not affected, so we don't really have a stake in the game.

@ChadKillingsworth ooc, how does NodeJS do ES6 imports of CJS code? Just `require` calls, or do they allow `import` statements?

**Comment by ChadKillingsworth** [comment]
@mprobst Here's a recentish writeup on the current state of things: http://2ality.com/2017/09/native-esm-node.html

One note: There is a very large developer backlash against the .mjs file extension. I'm kind of waiting to see how this plays out.

**Comment by mprobst** [comment]
Interesting. So NodeJS actually allows `import` statements to bring in non-ES-module code.

By the way, I forgot to say:

> But in either case you have no idea what your import statement is doing until you open the file you're importing due to the way CJS and goog.module handle default exports (exports = value). With ES6 modules import * is always an object that contains every export as a property; but this may not be the case with goog.module and CJS. Additionally what does a default import even mean for CJS/goog.module?

* having to read the imported file to figure out how it exports is a bit unfortunate, but given that we have default export/imports and named exports/imports, that's a reality I'm afraid
* we decided to represent `goog.module`-style `exports = ...;` as default exports in ES6 modules. That requires shimming the module object, but is IMHO the closest match for what the user intended.
* `import * as ...` just returns the `exports` object for `goog.module`, so `import * as x from 'goog:X'` is the same as `import x from 'goog:x';`.

Not saying these are the right choices, but they seem to hit a reasonably sweet spot regarding usability. They do generally require post-processing user source code and cannot be implemented as a library in JS though.

**Comment by ChadKillingsworth** [comment]
Here's the most recent counter proposal to Node's current direction: https://gist.github.com/ceejbot/b49f8789b2ab6b09548ccb72813a1054

**Comment by ribrdb** [comment]
I don't know if this is a use case you care about, but I'm using namespace imports extensively because Visual Studio Code understands them. I created a .d.ts file for the closure library and now I get intellisense and auto imports for closure library. If I switch to goog.require I would seriously miss this functionality.

**Comment by jplaisted** [comment]
This isn't proposing removing them from TypeScript, but from the Javascript compiler. So it depends on what the generated Javascript you're using looks like, assuming you're then using the Closure Compiler.

If you're going through Angular's tools / workflow (tsickle) then you should be all set. tsickle will rewrite these imports in the generated Javascript before the Closure Compiler runs.

**Comment by ribrdb** [comment]
I'm not using typescript. I'm using javascript with the bazel closure rules.

**Comment by ribrdb** [comment]
Just to clarify, I use the .d.ts to give visual studio code type information on the closure library. It's only used for intellisense. My code is just plain javascript.

**Comment by jplaisted** [comment]
In that case I think we can keep them around, but discourage them. And only have them work via the compiler so other google internal tools can stop caring about them.
